### PR TITLE
Training optimizations

### PR DESCRIPTION
**Pull request by david-ford**
Based on the changes made in [textual_inversion](https://github.com/rinongal/textual_inversion), I carried over the relevant changes that improved model training. These changes reduce the amount of memory used, significantly improve the speed at which training runs, and improve the quality of the results.

It also fixes the problem where the model trainer wouldn't automatically stop when it hit the set number of steps.

### CODE REVIEW

**Comment by blessedcoolant** [comment]
`RuntimeError: Distributed package doesn't have NCCL built in`

Dropping that error for me.

**Comment by blessedcoolant** [comment]
Sadly this OOM's for me even at 256. Same as the current implementation. Was hopeful that this update would let it run on my system but maybe not.

`sim = einsum('b i d, b j d -> b i j', q, k) * self.scale`

**Comment by lugia19** [comment]
> ~Sadly this OOM's for me even at 256 @ `sim = einsum('b i d, b j d -> b i j', q, k) * self.scale`~
> 
> Edit: Got it to work at 256 on an RTX 3080 8GB card. Nice job. Testing out the PR now.
> 
> The setting I had to change to get it to work is changing the size under training to 256. Everything else is set at 512. Wondering if that'll cause a conflict. I'll train this embed for an hour and see how it goes.

This is odd, I continue getting OOM despite being at 256 on an RTX 3080 10GB. Were there any other settings you changed besides the one on line 55?

**Comment by blessedcoolant** [comment]
> This is odd, I continue getting OOM despite being at 256 on an RTX 3080 10GB. Were there any other settings you changed besides the one on line 55?

I did not change 55. I changed 82.

**Comment by david-ford** [comment]
> 

The one that will have the biggest impact on your memory usage is line 76.

File: configs/stable-diffusion/v1-finetune.yaml [hunk 1]
```diff
@@ -73,7 +73,7 @@ model:
 data:
   target: main.DataModuleFromConfig
   params:
-    batch_size: 2
+    batch_size: 1
```
Comment (david-ford) [inline configs/stable-diffusion/v1-finetune.yaml:76]: This one right here.

**Review by blessedcoolant — APPROVED** [review]
Definitely optimizes it. I'd also recommend setting the `train` value to `256` as base instead of `512` because the majority of the people will have 8GB cards and I know this setting works for sure on 8GB cards.

But overall, I didn't notice any problems so far. Did 5 Epochs. Takes a while but stable.

**Review by lstein — APPROVED** [review]
Thanks for the contribution.

**Pull request merged**